core/types/calls.cc

```diff
@@ -962,6 +970,107 @@ DispatchResult dispatchCallSymbol(const GlobalState &gs, const DispatchArgs &arg
         implicitKwsplat = true;
     }
 
+    {
+        const ArgInfo &probablyFirstKwarg = *pit;
+        if (gs.ruby3KeywordArgs && args.args.size() > 0 &&
+            // All positional arguments are already processed
+            ait == aPosEnd &&
+            // But not all arguments are processed, some keyword arguments remain
+            ait != aend &&
+            // First keyword parameter doesn't look like keyword param
+            !probablyFirstKwarg.flags.isKeyword && !probablyFirstKwarg.flags.isRepeated &&
+            !probablyFirstKwarg.flags.isBlock) {
+            auto isLastParam /* not counting last block param */ =
+                (pit - data->arguments.begin()) >= data->arguments.size() - 2;
+
+            auto areOtherParamsDefault = !isLastParam;
+            auto curIdx = pit - data->arguments.begin();
+
+            for (int i = data->arguments.size() - 1; i > curIdx; i--) {
+                auto flags = data->arguments[i].flags;
+                if (flags.isBlock || flags.isRepeated) {
+                    continue;
+                }
+                if (!flags.isDefault) {
+                    areOtherParamsDefault = false;
+                    break;
+                }
+            }
+
+            auto maybeKwargIdx =
+                ait - args.args.begin() < args.args.size() ? ait - args.args.begin() : args.args.size() - 1;
+            auto allArgsProcessed = false;
+
+            // `hasKwsplat` can be false positive, because it relies on how we parse positional and keyword args
+            auto actuallyHasKwsplatParam = absl::c_any_of(
+                data->arguments, [](const auto &arg) { return arg.flags.isKeyword && arg.flags.isRepeated; });
```

Comment:
>I don't know whether it's correct to say that "hasKwsplat can be false positive"

It was tricky to figure out what `hasKwsplat` actually did. Firstly, because of the name confusion you've mentioned. I legitimately thought it was about param list, not an arg list.

Secondly, I noticed this:
```ruby
sig {params(a: Integer, h: T::Hash[T.untyped, T.untyped]).void}
def takes_empty_hash(a, h); end
empty_hash = {}
blk = Proc.new {puts "in proc"}

takes_empty_hash(1, **empty_hash) # => hasKwsplat = true
takes_empty_hash(1, **empty_hash, &blk) # => hasKwsplat = false
```

My expectation is `hasKwsplat` should be positive in both cases